flang/lib/Lower/SymbolMap.h

```diff
@@ -0,0 +1,239 @@
+struct SymIndex {
+  // For lookups that fail, have a monostate
+  using None = std::monostate;
+
+  // Capture triple notation ssa-values
+  using Triple = std::tuple<mlir::Value, mlir::Value, mlir::Value>;
+
+  // Trivial intrinsic type
+  struct Intrinsic {
+    explicit Intrinsic(mlir::Value addr) : addr{addr} {}
+    mlir::Value addr;
+  };
+
+  // Array variable that has a simple shape
+  struct Shaped {
+    explicit Shaped(mlir::Value addr, llvm::ArrayRef<mlir::Value> s)
+        : addr{addr}, shape{s.begin(), s.end()} {}
+    mlir::Value addr;
+    std::vector<mlir::Value> shape;
+  };
+
+  // Array variable that uses triple notation
+  struct FullDim {
+    explicit FullDim(mlir::Value addr, llvm::ArrayRef<Triple> s)
+        : addr{addr}, shape{s.begin(), s.end()} {}
+    mlir::Value addr;
+    std::vector<Triple> shape;
+  };
+
+  // CHARACTER type variable with its dependent type LEN parameter
+  struct Char {
+    explicit Char(mlir::Value addr, mlir::Value len) : addr{addr}, len{len} {}
+    mlir::Value addr;
+    mlir::Value len;
+  };
+
+  // CHARACTER array variable that has a simple shape
+  struct CharShaped {
+    explicit CharShaped(mlir::Value addr, mlir::Value len,
+                        llvm::ArrayRef<mlir::Value> s)
+        : addr{addr}, len{len}, shape{s.begin(), s.end()} {}
+    mlir::Value addr;
+    mlir::Value len;
+    std::vector<mlir::Value> shape;
+  };
+
+  // CHARACTER array variable using triple notation
+  struct CharFullDim {
+    explicit CharFullDim(mlir::Value addr, mlir::Value len,
+                         llvm::ArrayRef<Triple> s)
+        : addr{addr}, len{len}, shape{s.begin(), s.end()} {}
+    mlir::Value addr;
+    mlir::Value len;
+    std::vector<Triple> shape;
+  };
+
+  // Generalized derived type variable
+  struct Derived {
+    explicit Derived(mlir::Value addr, mlir::Value size,
+                     llvm::ArrayRef<Triple> s,
+                     llvm::ArrayRef<mlir::Value> parameters)
+        : addr{addr}, size{size}, shape{s.begin(), s.end()},
+          params{parameters.begin(), parameters.end()} {}
+    mlir::Value addr;
+    mlir::Value size;                // element size or null
+    std::vector<Triple> shape;       // empty for scalar
+    std::vector<mlir::Value> params; // LEN type parameters, if any
+  };
+
+  //===--------------------------------------------------------------------===//
+  // Constructors
+  //===--------------------------------------------------------------------===//
+
+  SymIndex() : v{None{}} {}
+  template <typename A>
+  SymIndex(const A &x) : v{x} {}
+
+  operator bool() const { return !std::holds_alternative<None>(v); }
+  operator mlir::Value() const { return getAddr(); }
+
+  //===--------------------------------------------------------------------===//
+  // Accessors
+  //===--------------------------------------------------------------------===//
+
+  mlir::Value getAddr() const {
```

Comment:
In case the symbol has a descriptor, is `getAddr()` the address of the descriptor or of the data ?

Reply:
It was meant to be the address of the data.